packages/menu/src/Menu.tsx

```diff
@@ -35,87 +52,114 @@ interface MenuProps extends Omit<PopoverProps, 'spacing'> {
 </button>
 ```
  * @param props.children Content to appear inside of Menu.
- * @param props.active Boolean to describe whether or not Menu is active.
+ * @param props.open Boolean to describe whether or not Menu is open.
+ * @param props.setOpen Callback to change the open state of the Menu.
+ * @param props.shouldClose Callback to determine whether or not Menu should close when user tries to close it.
  * @param props.className Classname applied to Menu.
  * @param props.align Alignment of Menu relative to another element: `top`, `bottom`, `left`, `right`.
  * @param props.justify Justification of Menu relative to another element: `start`, `middle`, `end`.
  * @param props.refEl Reference element that Menu should be positioned against.
  * @param props.usePortal Boolean to describe if content should be portaled to end of DOM, or appear in DOM tree.
  * @param props.trigger Trigger element can be ReactNode or function, and, if present, internally manages active state of Menu.
  */
+
+const escapeKey = 27;
+
 function Menu({
   align = Align.Bottom,
   justify = Justify.End,
   usePortal = true,
   adjustOnMutation = false,
-  active,
+  open = false,
+  shouldClose = () => true,
+  setOpen,
   children,
   className,
   refEl,
   trigger,
   ...rest
 }: MenuProps) {
-  const [isActive, setActiveState] = useState(false);
+  const [uncontrolledOpen, setUncontrolledOpen] = useState(false);
+
+  const popoverRef: React.RefObject<HTMLDivElement> = useRef(null);
+
+  const handleClose = () => {
+    if (setOpen && shouldClose()) {
+      setOpen(false);
+    }
+
+    if (!setOpen && shouldClose()) {
+      setUncontrolledOpen(false);
+    }
+  };
+
+  const handleBackdropClick = (e: MouseEvent) => {
+    const popoverReference = popoverRef && popoverRef.current;
+
+    if (
+      popoverReference &&
+      !popoverReference.contains(e.target as HTMLElement)
+    ) {
+      handleClose();
+    }
+  };
+
+  const enabled = open || uncontrolledOpen;
+
+  useEventListener('click', handleBackdropClick, {
+    enabled,
+  });
+
+  const handleEscape = (e: KeyboardEvent) => {
+    e.stopImmediatePropagation();
+
+    if (e.keyCode === escapeKey) {
+      handleClose();
+    }
+  };
+
+  useEventListener('keydown', handleEscape, {
+    enabled,
+  });
 
   const popoverContent = (
     <Popover
       key="popover"
-      active={trigger ? isActive : active}
+      active={setOpen ? open : uncontrolledOpen}
```

Comment:
Shouldn't we be able to do `open || uncontrolledOpen` here?

Reply:
👍